core/src/main/java/org/pentaho/di/core/logging/LoggingRegistry.java

```diff
@@ -22,130 +22,166 @@
+ * LoggingRegistry. It is a static singleton, and manages the in memory LoggingObjects as well as LoggingObjects that
+ * have FileWriter Buffers.
+ *
+ * The class provides a central and thread safe place to register, remove and find LoggingObjects held in memory.
+ *
+ * It uses a purgeTimer task to attempt to remove older registry objects as the internal map reaches the maxSize which
+ * is defined by the Kettle Property "KETTLE_MAX_LOGGING_REGISTRY_SIZE".
+ */
 public class LoggingRegistry {
+  /** private static reference **/
   private static LoggingRegistry registry = new LoggingRegistry();
+
+  /** Registry's LoggingObject Map containing objects keyed by object's LogChannelID **/
   private Map<String, LoggingObjectInterface> map;
+
+  /** Registry's FileWriterBuffer Map containing objects keyed by object's LogChannelID **/
   private Map<String, LogChannelFileWriterBuffer> fileWriterBuffers;
+
+  /** Map containing a list of LogChannelIds that belong toa parent Object. The key is the parent's LogChannelID **/
   private Map<String, List<String>> childrenMap;
+
+  /** Registry's PurgeQueue where Registry will remove entries based on FIFO performed on PurgeTimer Task **/
+  private Queue<LoggingObjectInterface> registerPurgeQueue;
+
+  /** Last time a Logging object was registered to the Registry **/
   private Date lastModificationTime;
-  private int maxSize;
-  private final int DEFAULT_MAX_SIZE = 10000;
 
+  /** Max Size of the Registry that it shall purge based on. Note: This is not a strict/hard limit, the Map object can
+   * grow above this number and will it attempt to "purge" below it. **/
+  private int maxSize;
+  private static final int DEFAULT_MAX_SIZE = 10000;
+
+  /** Timer object that executes Purge method based on purgeTimeout. **/
+  private Timer purgeTimer;
+  private int purgeTimeout;
+  private static final int DEFAULT_PURGE_TIMER = 60000;
+
+  // Statistics
+  /** Stat that counts the amount of the times the purge task is invoked.**/
+  private int purgeTimerCount;
+  /** Stat that counts the amount of Objects removed from registry map.**/
+  private int purgedObjectCount;
+  /** Stat that counts how many objects are found instead of added to the map.**/
+  private int foundCounter;
+
+  /** Sync object **/
   private final Object syncObject = new Object();
 
   private LoggingRegistry() {
     this.map = new ConcurrentHashMap<>();
     this.childrenMap = new ConcurrentHashMap<>();
     this.fileWriterBuffers = new ConcurrentHashMap<>();
+    this.registerPurgeQueue = new ConcurrentLinkedQueue<>();
 
     this.lastModificationTime = new Date();
     this.maxSize = Const.toInt( EnvUtil.getSystemProperty( "KETTLE_MAX_LOGGING_REGISTRY_SIZE" ), DEFAULT_MAX_SIZE );
+    this.purgeTimeout = Const.toInt( EnvUtil.getSystemProperty( "KETTLE_LOGGING_REGISTRY_PURGE_TIMEOUT" ), DEFAULT_PURGE_TIMER );
+    this.purgeTimerCount = 0;
```

Comment:
KETTLE_LOGGING_REGISTRY_PURGE_TIMEOUT should probably be a constant in Const.java for consistency with most other kettle params.  I know KETTLE_MAX_LOGGING_REGISTRY_SIZE was already there before, but should probably also be defined in Const.java.

Reply:
Fixed, I also resolved https://jira.pentaho.com/browse/PDI-18341 in the latest force push.